### ISSUE TITLE
Instability in gpu_hist

### ISSUE THREAD

**Issue by pseudotensor** [dmlc/xgboost#3156]
```
import xgboost
import pickle
import numpy as np
pathpickle='c92b4c2f-38c3-47ba-94f9-95af4058dcad_xgbobj2.dat'

model, XX_pd, y, _, kwargs = pickle.load(open(pathpickle, 'rb'))
del kwargs['sample_weight_eval_set']
del kwargs['early_stopping_threshold']
del kwargs['early_stopping_limit']
#del kwargs['early_stopping_rounds']
#kwargs['early_stopping_rounds'] = 30
#kwargs['eval_metric'] = "logloss"
#del kwargs['eval_set']
#kwargs['eval_set'] = (kwargs['eval_set'][0], (XX_pd, y))
kwargs['eval_set'] = ((XX_pd, y),)
#kwargs['verbose'] = False
#model.kwargs['tree_method'] = 'hist'
model.fit(XX_pd, y, **kwargs)
#model.fit(XX_pd, y)
```

[c92b4c2f-38c3-47ba-94f9-95af4058dcad_xgbobj2.zip](https://github.com/dmlc/xgboost/files/1791795/c92b4c2f-38c3-47ba-94f9-95af4058dcad_xgbobj2.zip)

If any type of eval set is passed, even the training set as above, still happens.  The training score does not descrease monotonically.  Using "hist" works correctly.

**Comment by pseudotensor** [comment]
@RAMitchell

**Comment by RAMitchell** [comment]
Fixed this bug in #3167

Turns out both hist and gpu_hist were incorrect.
